aerogear/aerogear-unifiedpush-server
# Aerogear 9672 - Create a retry/queuing mechanism for JMS messages

## Motivation
We're currently using enmasse as one of our supported messaging systems.  However, enmasse does not support (currently) DLQ/retry configuration.  This PR let's UPS control requeuing those jms messages when we would have previously relied on the message broker.

## What
This PR wraps the notification dispatcher in a try/catch block that will handle run time exceptions by looking at new metadata in the JMS message and optionally requeuing it.  This behavior is configurable through environment variables.

This PR adds checks two environment variables : AMQ_MAX_RETRIES, and AMQ_BACKOFF_SECONDS.  AMQ_MAX_RETRIES is the number of retries that will be sent.*  AMQ_BACKOFF_SECONDS is the number of seconds to increase the delay timeout.  


 * behavior note, this means that if you have a failure and three retries that fail you will see four error messages.  One for the original and three for the retries.

## Verification Steps
Add the steps required to check this change. Following an example.
 
1. Configure a push application normally
2. confirm it works
3. Lobotomize the push server somehow; (block the connection to the external push services, set up a bad iOS certificate, add a throws runtime exception to a sender and rebuild/redeploy with that, etc 
 The point is it has to be really rude and really bad)
4. Send a push notification and confirm that your messages are being requeued
